src/ctap2/preflight.rs

```diff
@@ -0,0 +1,211 @@
+        let key_handle = self
+            .get_credential_id_list()
+            .iter()
+            // key-handles in CTAP1 are limited to 255 bytes, but are not limited in CTAP2.
+            // Filter out key-handles that are too long (can happen if this is a CTAP2-request,
+            // but the token only speaks CTAP1).
+            .filter(|key_handle| key_handle.id.len() < 256)
+            .find_map(|key_handle| {
+                let check_command = CheckKeyHandle {
+                    key_handle: key_handle.id.as_ref(),
+                    client_data_hash: self.get_client_data_hash().as_ref(),
+                    rp: self.get_rp(),
+                };
+                let res = dev.send_ctap1(&check_command);
+                match res {
+                    Ok(_) => Some(key_handle.clone()),
+                    _ => None,
+                }
+            });
+        self.handle_ctap1_pre_flight_key(dev, key_handle)
+    }
+
+    fn do_pre_flight_ctap2<Dev: FidoDevice>(
+        &mut self,
+        chunk_size: usize,
+        dev: &mut Dev,
+    ) -> Result<(), AuthenticatorError> {
+        // If there is max. one entry, we can skip all of this
+        if self.get_credential_id_list().len() <= 1 {
+            return Ok(());
+        }
+
+        let original_list = self.get_credential_id_list().to_vec();
+        let chunked_list = original_list.chunks(chunk_size);
+        // Step 1.2: Return early, if we only have one chunk anyways
+        if chunked_list.len() <= 1 {
+            return Ok(());
+        }
+
+        // Step 2: If we have more than one chunk: Loop over all, doing GetAssertion
+        //         and if one of them comes back with a success, use only that chunk.
+        self.set_credential_id_list(vec![]);
+        for chunk in chunked_list {
+            let mut silent_assert = GetAssertion::new(
+                ClientDataHash(Sha256::digest("").into()),
+                self.get_rp().clone(),
+                chunk.to_vec(),
+                GetAssertionOptions {
+                    user_verification: if self.get_pin_uv_auth_param().is_some() {
+                        None
+                    } else {
+                        Some(false)
+                    },
+                    user_presence: Some(false),
+                },
+                GetAssertionExtensions::default(),
+                None,
+                None,
+            );
+            silent_assert.set_pin_uv_auth_param(
+                self.get_pin_uv_auth_param()
+                    .map(|p| p.pin_auth_token.clone()),
+            )?;
+            let res = dev.send_msg(&silent_assert); // TODO(MS): Cancellable!
+            match res {
+                Ok(..) => {
+                    // This chunk contains a key_handle that is already known to the device.
+                    // Replace credential_id_list with current chunk.
+                    self.set_credential_id_list(chunk.to_vec());
```

Comment:
I don't think this is right for `GetAssertion`. An allow list might have a number of valid credentials that exceeds the chunk size. This will not return all of them. Changing this to append the chunk won't work because then the final request won't fit in a single chunk.

Reply:
I was convinced by your message on matrix that this is OK. The spec does not require us to present all of the credentials in the allow list to the authenticator.